pkg/controller/nodemaintenance/drain.go

```diff
@@ -0,0 +1,190 @@
+package nodemaintenance
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/kubectl/drain"
+)
+
+func runCordonOrUncordon(r *ReconcileNodeMaintenance, node *corev1.Node, desired bool) error {
+	cordonOrUncordon := "cordon"
+	if !desired {
+		cordonOrUncordon = "un" + cordonOrUncordon
+	}
+
+	log.Info(fmt.Sprintf("%s Node: %s", cordonOrUncordon, node.Name))
+
+	c := drain.NewCordonHelper(node)
+	if updateRequired := c.UpdateIfRequired(desired); updateRequired {
+		err, patchErr := c.PatchOrReplace(r.drainer.Client)
+		if patchErr != nil {
+			log.Error(err, fmt.Sprintf("Unable to %s Node %s \n", cordonOrUncordon, node.Name))
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func drainPods(r *ReconcileNodeMaintenance, nodeName string) error {
+	list, errs := r.drainer.GetPodsForDeletion(nodeName)
+
+	if errs != nil {
+		return utilerrors.NewAggregate(errs)
+	}
+
+	if warnings := list.Warnings(); warnings != "" {
+		log.Info(fmt.Sprintf("WARNING: %s\n", warnings))
+	}
+
+	if err := deleteOrEvictPods(r, list.Pods()); err != nil {
+		pendingList, newErrs := r.drainer.GetPodsForDeletion(nodeName)
+		log.Error(err, fmt.Sprintf("There are pending pods in node %q when an error occurred: \n", nodeName))
+
+		for _, pendingPod := range pendingList.Pods() {
+			log.Error(err, fmt.Sprintf("%s/%s\n", "pod", pendingPod.Name))
+		}
+		if newErrs != nil {
+			log.Error(err, fmt.Sprintf("following errors also occurred:\n%s", utilerrors.NewAggregate(newErrs)))
+		}
+		return err
+	}
+	return nil
+}
+
+// deleteOrEvictPods deletes or evicts the pods on the api server
+func deleteOrEvictPods(r *ReconcileNodeMaintenance, pods []corev1.Pod) error {
+	if len(pods) == 0 {
+		return nil
+	}
+
+	policyGroupVersion, err := drain.CheckEvictionSupport(r.drainer.Client)
+	if err != nil {
+		return err
+	}
+
+	getPodFn := func(name string) (bool, error) {
+		_, exists, err := r.podInformer.GetStore().GetByKey(name)
```

Comment:
Here you have to pass `namespace + "/" + name`, otherwise you will not be able to find the pod.